[1.21 dev] bunch of mostly poorly done 1.21 work

A5ho9999 opened the pull request:
This should complete a lot of the 1.21 work mainly the data components and enchantment stuff (it's a working 1.21.1 version as far as I can tell)

I'm not sure about the reputation and pricing systems but they seem correct

I didn't know what path you guys want to take for the Data Components so I've just applied it for the custom data type until then.

Feel free to scrap this whole thing if it's not any good

A5ho9999 committed:
Some 1.21 work

A5ho9999 committed:
Some 1.21 work

A5ho9999 committed:
NBT stuff

A5ho9999 committed:
smol enchant fix

A5ho9999 committed:
Fix CopyNBT, ShopOffer Endec, TradePacket and Loot-tables

A5ho9999 committed:
Fix randomized enchants

A5ho9999 committed:
Include the double price tag

Noaaan reviewed (changes requested):
This PR seems fairly good. Need to do some more rigorous testing, but overall it is a nice baseline.

Only thing I want changed is checkstyle. The Wisp Forest codebases, including Numismatic, mainly use `var` instead of explicit type declarations. 
```java
// Normal declaration
Object object = new Object();
Object object2 = someComplexFunction();

// Using var
var object = new Object();
var object2 = someComplexFunction();
```

A5ho9999 committed:
correct style
fix error with modded villagers in RemappingTradeWrapper
fix purse not removing coins when trading

A5ho9999 commented:
Switched those that could be var, was one that couldn't be changed. Any others weren't done by me.

Also corrected 2 bugs, one being pretty obvious that I don't know how missed it. The other relating to the [remapping](https://gist.github.com/A5ho9999/e2c0dbd3e1d053c099fe4b58c3fa4286) which was also mentioned on the discord.

Noaaan commented on src/main/java/com/glisco/numismaticoverhaul/villagers/json/adapters/EnchantItemAdapter.java:
```diff
@@ -55,18 +64,93 @@ public Factory(ItemStack item, int maxUses, int experience, int level, boolean a
         }
 
+
+        // TODO: Review
```
This does not seem right...

![image](https://github.com/user-attachments/assets/534b8fe8-18f6-498b-8a9c-485e39059ce3)

I do need to spend some time reviewing the villager trading changes again. From what I remember the following applies:

- A Villager never enchants an Item with a Treasure enchantment
- Books can contain treasure enchantments

An irrelevant sidenote for this port (which I am writing so I won't forget), is that 1.21.5+ changed trades from being randomly generated to a set list dependent on the Villagers biome. This is fortunately not important for this port, but we want to keep vanilla (1.21-like) behavior here. 

Can you not just use the EnchantmentHelper method like before? It just needs the `tradeRegistry`

A5ho9999 replied:
I did overlook the equipment and only checked the enchanted books, so that's on me. I'll have to look into how to rework this better.

As for the 1.21.5 change, that's not the case as far as I'm aware. That part still remains experiential and the only trade related change is the cartographer selling maps to other villagers unless I'm missing something.

The EnchantmentHelper methods have changed a lot and are rather restrictive imo. It won't accept the `tradeRegistry` as is. I can make it work but it would no longer support the allowTreasure option. 

I'm also not sure which items are handled in this class. If it's for both Equipment and Books then it would need some major reworking as Equipment and Books are completely different enchantment tags.

To change to using the EnchantmentHelper:
```
var dynamicRegistry = entity.getWorld().getRegistryManager();
            
EnchantmentHelper.enchant(random, itemStack, level, tradeRegistry.map(RegistryEntryList::stream).orElseGet(() -> dynamicRegistry.get(RegistryKeys.ENCHANTMENT).streamEntries().map(enchantmentReference -> enchantmentReference)));
```

Noaaan replied:
I have gone ahead and changed the adapter to more closely match the original method. Refer to [d50fc34](https://github.com/wisp-forest/numismatic-overhaul/pull/157/commits/d50fc34a9077039da4aea4b811478597fdcb6e38)

Noaaan replied:
As for the 1.21.5 changes - yeah, this is not something to prioritize in this PR.

Noaaan commented on src/main/java/com/glisco/numismaticoverhaul/mixin/SetTradeOffersS2CPacketMixin.java:
```diff
@@ -0,0 +1,42 @@
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+import org.spongepowered.asm.mixin.injection.callback.LocalCapture;
+
+// TODO: Review, not sure if this is correct. Seems to work?
```
Reputation seems to work as intended. Note that this exists in order to combat mass-curing Villagers for essentially free items (coins are more volatile to this than Emeralds)

A5ho9999 replied:
if it's for the mass curing would it still be needed? I don't remember which version but I know in one of the 1.21s they removed the ability to cure more than once.

Noaaan replied:
Yes, this is correct. I think it is best to fully remove this now, as it should allow for better compat overall (as long as it works)

Noaaan committed:
Fix single-coin tooltips, add some TODOs

Noaaan committed:
Rework EnchantItemAdapter

Use EnchantmentHelper and specific registry lists,
which removes the need for the fallback.

A5ho9999 committed:
Remove no longer needed Reputation system
